packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,261 @@
+import { NodePHP } from '@php-wasm/node';
+
+let childCount = 0;
+
+/**
+ * PRIVATE
+ * Tracks stats of instances in a pool.
+ */
+class PoolInfo {
+	id = childCount++; // Unique ID for debugging purposes.
+	requests = 0; // Total requests processed.
+	started = Date.now(); // Time spawned.
+	active = false; // Whether instance is considered active.
+}
+
+/**
+ * PRIVATE
+ * Spawns new instances if the pool is not full.
+ * Returns a list of new instances.
+ */
+const spawn = (pool) => {
+	const newInstances = new Set();
+
+	while (pool.maxJobs > 0 && pool.instanceInfo.size < pool.maxJobs) {
+		const info = new PoolInfo();
+		const instance = pool.spawner();
+		pool.instanceInfo.set(instance, info);
+		info.active = true;
+		newInstances.add(instance);
+	}
+
+	return newInstances;
+};
+
+/**
+ * PRIVATE
+ * Reaps children if they've passed the maxRequest count.
+ */
+const reap = (pool) => {
+	for (const [instance, info] of pool.instanceInfo) {
+		if (pool.maxRequests > 0 && info.requests >= pool.maxRequests) {
+			info.active = false;
+			pool.instanceInfo.delete(instance);
+			instance.then((unwrapped) => unwrapped.exit()).catch((error) => {});
+			continue;
+		}
+	}
+};
+
+/**
+ * PRIVATE
+ * Handle fatal errors gracefully.
+ */
+const fatal = (pool, instance, error) => {
+	console.error(error);
+
+	if (pool.instanceInfo.has(instance)) {
+		const info = pool.instanceInfo.get(instance);
+		info.active = false;
+		pool.instanceInfo.delete(instance);
+	}
+};
+
+/**
+ * PRIVATE
+ * Find the next available idle instance.
+ */
+const getIdleInstance = (pool) => {
+	const sorted = [...pool.instanceInfo].sort(
+		(a, b) => a[1].requests - b[1].requests
+	);
+
+	for (const [instance, info] of sorted) {
+		if (pool.running.has(instance)) {
+			continue;
+		}
+
+		if (!info.active) {
+			continue;
+		}
+		return instance;
+	}
+
+	return false;
+};
+
+/**
+ * Maintains and refreshes a list of php instances
+ * such that each one will only be fed X number of requests
+ * before being discarded and replaced.
+ */
+export class Pool {
+	instanceInfo = new Map(); // php => PoolInfo
+
+	spawner: () => Promise<any>; // Callback to create new instances.
+	maxRequests: number; // Max requests to feed each instance
+	maxJobs: number; // Max number of instances to maintain at once.
+
+	notifiers = new Map(); // Inverted promises to notify async code of backlogged item processed.
```

Comment:
Nit: I had to check with `enqueue` what is this doing as I didn't understand the name and the comment. Perhaps `resolvers` would be more descriptibe, since `new Promise(resolve => { ... });` is a popular pattern 🤔

Reply:
@adamziel Yea, this is definitely a weird pattern. I've only seen it implemented in a handful of places in my career. I like the name `resolvers` for this field. Its definitely more descriptive.

Reply:
Thanks!

I saw a name `PromiseQueue` used for this general pattern. There was even a generic implementation used in Playground some months back but it was removed and simplified later on.